lib/remote/jsonrpcconnection-pki.cpp

```diff
@@ -78,12 +78,45 @@ Value RequestCertificateHandler(const MessageOrigin::Ptr& origin, const Dictiona
 	}
 
 	if (signedByCA) {
-		if (IsCertUptodate(cert)) {
+		bool uptodate = IsCertUptodate(cert);
 
+		if (uptodate) {
+			// Even if the leaf is up-to-date, the root may expire soon.
+			// In a regular setup where Icinga manages the PKI, there is only one CA.
+			// Icinga includes it in handshakes, let's see whether the peer needs a fresh one...
+
+			// The following analysis targets a direct peer's cert chain and makes no sense for forwarded CSRs.
+			if (cn == origin->FromClient->GetIdentity()) {
+				auto chain (SSL_get_peer_cert_chain(tlsConn.native_handle()));
+
+				if (chain) {
+					X509* root = nullptr;
+					auto len (sk_X509_num(chain));
+
+					for (int i = 0; i < len; ++i) {
+						auto link (sk_X509_value(chain, i));
+
+						if (!X509_NAME_cmp(X509_get_subject_name(link), X509_get_issuer_name(link))) {
+							root = link;
+						}
+					}
+
+					if (root) {
+						uptodate = IsCertUptodate(root);
+					}
```

Comment:
What do you worry about? Another reconnect loop? Like the master's CA just became !IsCertUptodate(), but the RENEW_INTERVAL isn't due soon? And the nodes below fetch the old CA over and over again? Then this may be not a proper fix as we don't know the number of cluster levels. My previous approach was to let satellites directly serve renewal requests if they have a newer CA. The current approach is to forward it up ef50e5ac2 and finally handle it on the master b43f1e770. What about the satellite forwarding it up, but only if it (the satellite!) has a newer CA?

Reply:
> What do you worry about? Another reconnect loop? Like the master's CA just became !IsCertUptodate(), but the RENEW_INTERVAL isn't due soon? And the nodes below fetch the old CA over and over again?

Yes, exactly.

> Then this may be not a proper fix as we don't know the number of cluster levels.

I don't see how this depends on the number of cluster levels.

---

Let's take a step back and have a look at the overall picture. If the CA is renewed early enough (`LEAF_VALID_FOR` before it would expire), there's actually no need to perform an extraordinary renewal of the leaf certificate: it will expire before the CA and it will be renewed early enough which deploys the renewed/extended CA certificate.

So there's only one case that needs some special handling: the Icinga update bringing CA renewal was installed after leaf certificates were already issued that exceeded the validity of the CA at this time. But that situation is fairly easy to detect.

Here is what I would do:
1. Renew the CA early enough so that we can always issue certificates that are valid for `LEAF_VALID_FOR`.
2. When receiving a certificate request from a directly connected endpoint, test whether the leaf validity exceeds the CA validity, if so, add a flag to the certificate request, similar to how it's done with `ticket`.
3. When a CA signing master receives a certificate request that expires within `RENEW_THRESHOLD` or has that flag set, renew the certificate.
4. Add some safeguards to prevent reissuing loops (for example if `/var/lib/icinga2/ca/` is not writable): only reissue due to CA expiry if a CA cert is available that allows the full `LEAF_VALID_FOR` validity.

We could also consider limiting the validity of newly issued leaf certificates to `min(now + LEAF_VALID_FOR, ca_expiry)`, as issuing with longer validity makes no sense, but I think that would only matter if the CA renewal would fail and might require extra checks there to prevent reissuing loops (something like only reissue if it extends the validity), so not sure if that would be worth it.

Reply:
Fixed that loop in b7b448a4c (3ea99ad67 makes the difference). Now the master just says: `[2023-12-08 11:20:37 +0000] information/JsonRpcConnection: The certificates for CN 'sat' and its root CA are valid and uptodate. Skipping automated renewal.`

And if I restart the master to enforce the missed self-renewal and the agent for triggering a new CSR, everyone gets updated of course. 👍